### ISSUE TITLE
Compiler doesn't catch invalid mutable variable of `let mut var = &self.var` when it is `&self` instead of `&mut self` in method parameter

### ISSUE THREAD

**Issue by kw7oe** [rust-lang/rust#96041]
I tried this code:

```rust
pub struct DiskManager {
    file: File,
}

impl DiskManager {
    pub fn read(&self, offset: usize) -> Result<[u8; 4], std::io::Error> {
        // Take a reference to self.file, and declaring file as mutable.
        let mut file = &self.file;
        
        // Both file.seek and file.read_exact expect a &mut self.
        // But file is a `let mut file: &File`. 
        // 
        // So it's this actually an acceptable behaviour or a bug in the compiler?
        file.seek(SeekFrom::Start(offset as u64))?;
        let mut buffer = [0; 4];
        file.read_exact(&mut buffer)?;
        Ok(buffer)
    }
    
    // Shouldn't it be something like this?
    pub fn read(&mut self, offset: usize) -> Result<[u8; 4], std::io::Error> {
        self.file.seek(SeekFrom::Start(offset as u64))?;
        let mut buffer = [0; 4];
        self.file.read_exact(&mut buffer)?;
        Ok(buffer)
    }
}

```

I expected that the rust compiler will complain about `let mut file = &self.file` is not a mutable references. It should show something like:

```
 cannot borrow `self.file` as mutable, as it is behind a `&` reference
```

Instead, this code actually compiled successfully.

### Meta

Here's a link to Rust Playground for the demo code: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b3c0a34ff2b77bb22ca4fd1ca7ce89db

#### A bit more context

I was trying to implement concurrent access to read file. To have it thread safe, this means that my `read` function shouldn't take `mut self`. Hence, the correct way to do this would be through interior mutability. 

But when I was trying out the approach above to remove the need of `mut self`, it compiled. However, if this impl is being accessed concurrently, it will have undefined behaviour.

I'm fairly new to Rust, it might be not a bug but my lack of understanding about Rust.

**Comment by hellow554** [comment]
The variable `file` is mutable, not `self.file`. What you can't do is calling a method that takes `&mut self` on `file` it self. Let me showcase it to you:

```rust
fn main() {
    let a = String::from("abc");
    let mut b = &a;
    b.pop();
}
```

does not compile.

**Comment by hellow554** [comment]
The reason why this works is that the `Read` trait is implemented for `&File`: https://doc.rust-lang.org/std/fs/struct.File.html#impl-Read-1

If you change your file to a `io::Cursor` for example, your code will no longer work: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e6d1da79240e5398d597e5eeae831437

**Comment by MomoLangenstein** [comment]
I think this behaviour is intended as both `Seek` and `Read` are implemented for `File` **and** `&File`.

**Comment by kw7oe** [comment]
> The reason why this works is that the `Read` trait is implemented for `&File`: https://doc.rust-lang.org/std/fs/struct.File.html#impl-Read-1
> 
> If you change your file to a `io::Cursor` for example, your code will no longer work: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e6d1da79240e5398d597e5eeae831437
> 
> I mean, that's suspicious, but probably right 🤔

> I think this behaviour is intended as both Seek and Read are implemented for File and &File.

Ahh okay, now it make sense why it works. I think for me the confusing part is when I do:

```rust
pub fn read(&self, offset: usize) -> Result<[u8; 4], std::io::Error> {
  self.file.seek(SeekFrom::Start(offset as u64))?;
  let mut buffer = [0; 4];
  self.file.read_exact(&mut buffer)?;
  Ok(buffer)
}
```

I get the `cannot borrow as mutable` error, hence I assumed that it have to be mutable references. 

Since we declared `file` as `mut`, when we call `file.seek`, it's still correct for the `&File.read(&mut self, ...)` interface. So it would only failed if the trait is implemented on `File` only. And since it run out of scope after the function ended, hence borrow checker is not complaining as well.

And the reason `self.file.seek` doesn't work without declaring it as `mut` is because we are operating on `File`. So regardless of `File` or `&File`, both take `&mut self`. And hence why the error occur for that case.

So I guess, I'm getting a bit confused by the type due to the lack of understanding of it. 

Having said that, what would be the right way to actually implement it though? Is both the equally good?

Please let me know if I should move the question to another channel.

**Comment by hellow554** [comment]
You can ask your questions everytime at https://users.rust-lang.org :heart:
